src/System.Buffers.ReaderWriter/System/Buffers/Reader/BufferReader_binary.cs

```diff
@@ -2,52 +2,167 @@
 // Licensed under the MIT license. See LICENSE file in the project root for full license information.
 
 using System.Buffers.Binary;
+using System.Diagnostics;
+using System.Runtime.InteropServices;
 
 namespace System.Buffers.Reader
 {
-    public static partial class BufferReaderExtensions
+    public ref partial struct BufferReader
     {
-        public static bool TryRead(ref BufferReader reader, out int value, bool littleEndian = false)
+        /// <summary>
+        /// Try to read the given type out of the buffer if possible.
+        /// </summary>
+        /// <remarks>
+        /// The read is unaligned.
+        /// </remarks>
+        /// <returns>
+        /// True if successful. <paramref name="value"/> will be default if failed.
+        /// </returns>
+        public unsafe bool TryRead<T>(out T value) where T : unmanaged
         {
-            var unread = reader.UnreadSegment;
-            if (littleEndian)
+            ReadOnlySpan<byte> span = UnreadSegment;
+            if (span.Length < sizeof(T))
+                return TryReadSlow(out value);
+
+            value = MemoryMarshal.Read<T>(span);
+            Advance(sizeof(T));
+            return true;
+        }
+
+        private unsafe bool TryReadSlow<T>(out T value) where T : unmanaged
+        {
+            Debug.Assert(UnreadSegment.Length < sizeof(T));
+
+            // Not enough data in the current segment, try to peek for the data we need.
+            byte* buffer = stackalloc byte[sizeof(T)];
```

Comment:
Why isn't this directly using stackalloc Span?

Reply:
Because it won't let you pass it to Peek (for fear of being captured as the reader is a ref struct). Readonly methods should address some of this. I'm discussing with @jaredpar etc. about the possibilities of adding a feature that allows methods to take spans and specify that they won't stash the incoming span argument.